### ISSUE TITLE
Repeat without out_dim causes BatchDim Error

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#200]
"Und täglich grüßt das Murmeltier" :sweat_smile: 

Not setting `out_dim` in the `repeat` layer causes the batch information of the created Dim to not be updated. I would assume this would also happen to other layers that create Dims, so I think an update in either `_generate_layers` or some other part of the code is required, thats why before attempting a fix I would like to discuss options with you.

```python
def test_repeat_without_out_dim():
  # https://github.com/rwth-i6/returnn_common/issues/200
  from returnn.config import Config
  from returnn.tf.engine import Engine
  from returnn.datasets import init_dataset
  time_dim = nn.SpatialDim("time")
  in_dim = nn.FeatureDim("in", 3)
  out_dim = nn.FeatureDim("out", 5)
  x = nn.Data("data", dim_tags=[nn.batch_dim, time_dim, in_dim], available_for_inference=True)


  def _config_get_network(epoch: int, **_kwargs) -> dict:
    # noinspection PyStatementEffect
    epoch  # unused
    nn.reset_default_root_name_ctx()
    net = nn.Linear(out_dim)
    y = net(nn.get_extern_data(x))
    out, dim = nn.repeat(y, repetitions=2, axis=time_dim)
    out.mark_as_default_output()
    y.mark_as_loss()
    net_dict = nn.get_returnn_config().get_net_dict_raw_dict(nn.Module())
    return net_dict

  config = Config({
    "task": "train", "num_epochs": 1, "start_epoch": 1,
    "get_network": _config_get_network,
    "extern_data": {x.name: {"dim_tags": [nn.batch_dim, time_dim, in_dim], "available_for_inference": True}},
  })
  train_dataset = init_dataset(
    {"class": "DummyDataset", "input_dim": in_dim.dimension, "output_dim": 5, "num_seqs": 3})
  engine = Engine(config)
  engine.init_train_from_config(config, train_data=train_dataset)
  engine.train()
```

```
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn/tf/network.py", line 183, in ExternData.init_batch_info
    line: assert not tag.dyn_size_ext.batch or tag.dyn_size_ext.batch == batch_info
    locals:
      tag = <local> Dim{':repeat:out_dim'[B(-1)]}
      tag.dyn_size_ext = <local> Data{'data_default_dyn_size_ext', [B(-1)], dtype='int32'}
      tag.dyn_size_ext.batch = <local> BatchInfo{B(-1)}
      batch_info = <local> BatchInfo{B}
AssertionError
```

The weird loss definition is intended, I will open another issue right after this.

**Comment by Atticus1806** [comment]
It does not really matter. I just found it easier since replicating the "working" version then allows for: 
`out_dim=time_dim` but you could also put e.g. 3 and then `out_dim=time_dim*3`

**Comment by Atticus1806** [comment]
In my code I use it with a tensor instead of int, for the simple example I found this sufficient since the issue is the same.

**Comment by albertz** [comment]
One problem is then here (in `RepeatLayer`), due to your special case `repetitions=1`:

```python
    if out_dim:
      out_dim_.declare_same_as(out_dim)
```
`out_dim_` is the same as `time_dim` (`Dim{'time'[B(-1)]}`) due to this, and this gets declared the same as the given `out_dim`. This is only a problem because `out_dim_` already exists, due to `repetitions=1`, although there could be other cases where it would exist as well. It's kind of the reverse problem we had before (https://github.com/rwth-i6/returnn/issues/1102).

So, we do not want that `declare_same_as` overwrites an existing dim with such a new dim.

I wonder about a generic way to cover such cases. Maybe we can track what depends on what (we already do), or what was created first.

**Comment by albertz** [comment]
Note that `dim_left.declare_same_as(dim_right)` same does multiple things:
* It has the effect that `dim_left == dim_right` afterwards. So it basically merges them.
* If `dim_right` is actually derived from `dim_left`, it performs the reverse operation.
* It overtakes the `dyn_size`/`dyn_size_ext`. In case both are defined, it overtakes it from `dim_right`. Otherwise it takes it where it gets it from.
* It overtakes the `description`. This is purely cosmetic, but it maybe matters for nicer debug output, and also it influences the variable names of dumped Python code for the dim tags.

It's actually not so clear what description to prefer in case of dim tag math. Do you like `time*2` more, or `repeat:out_dim`?

**Comment by albertz** [comment]
> It does not really matter.

It does. When you put `repetitions=2`, it causes a different problem.

**Comment by Atticus1806** [comment]
> It's actually not so clear what description to prefer in case of dim tag math. Do you like `time*2` more, or `repeat:out_dim`?

I would prefer `time*2`if the decision is split between those two, even though I would imagine something like: `repead:time*2` would be the best, so layer which creates and then what "value" is in it.

Okay I think I understand what you mean now, then lets go with `repetitions=2` here for now. Because that is closer related to my problem.

**Comment by albertz** [comment]
Wow, debugging this is really annoying.

It looked like `out_dim_.dyn_size_ext` somehow was inconsistent. Its batch dim was B while `out_dim_.batch` was None. I think this causes problems later. I'm still investigating this.

However, I wondered, where does it actually set the `out_dim_.dyn_size_ext`. I had a breakpoint here:
```python
out_dim_ = tag * repetitions
```
`tag` had it correct (consistent).

Stepping into this, into the constructor of the new `Dim`, there we actually had `dyn_size_ext = None`. **However**, in the debugger, in `Dim.__init__`, after the line:
```python
self.dyn_size_ext = dyn_size_ext
```
**Suddenly** `self.dyn_size_ext` was not None anymore? How can that be?

Even more strange, in the debugger console, after setting `self.dyn_size_ext = None` explicitly, then when checking it again, `self.dyn_size_ext` was not None anymore.

After debugging further, I found the reason:

PyCharm will evaluate the attribs of the objects to visualize them nicely. For that, it will potentially call Python code. Obviously there is e.g. `Dim.__repr__`, but that was not having any such side effects (good -- that would be bad). However, when you check `Dim.dyn_size`, this is a property, and the getter function has the side effect of setting `dyn_size_ext` when it is not yet set but when it can be inferred e.g. from dim tag math! So that explains it. It seems that it re-evaluates the object attribs by some heuristics, potentially after every execution line, to update the PyCharm GUI.

I'm not really sure if that all matters now for the issue here. But this got me quite distracted and confused, so I wanted to share this fun little story.

**Comment by albertz** [comment]
There are actually two maybe independent issues:

* [ ] The line `out_dim_.declare_same_as(out_dim)` is conceptually correct (see discussion in https://github.com/rwth-i6/returnn/issues/1102). However, in this case, `out_dim_` already exists, and should take priority over `out_dim`. This is what I mentioned above already. This needs to be well defined and then implemented.
* [ ] Even without the correct priority, why does this fail? I think it should not. This should be investigated.

**Comment by albertz** [comment]
> This needs to be well defined and then implemented.

This is not really so clear. Say we do `dim_left.declare_same_as(dim_right)`. There is already special handling if one is derived from the other, but assume we do not have this case here. The standard behavior then is to overtake `description` from `dim_right` (and other things, as explained above).

We have the case here in this issue for `RepeatLayer`:

* `dim_right` is undefined (given as `out_dim`, i.e. supposed to be defined) (actually, only undefined in the first call, and for the main net construction, it is defined but for the wrong batch info... -> maybe say undefined for the current batch info, or maybe don't care, as in further calls, it should then not matter anymore)
* `dim_left` is defined, and comes from extern_data.

Potential rules:

* If `dim_right` is undefined, `dim_left` is defined -> prioritize `dim_left`.
This rule has the problem that in many cases, `dim_left` is defined, often via dim tag math (e.g. pool/conv), and `dim_right` undefined, but here we might want to overtake `dim_right`, at least the description, at least in some cases?

* If `dim_right` is undefined, `dim_left` is defined and (its `description` comes from the user, or from extern_data) -> prioritize `dim_left`.
This rule seems to behave as we want. However, it looks a bit complicated, and it's not really clear how we check for that, or it likely needs some further implementation for that.

* If `dim_right` is undefined, `dim_left` is defined not using dim tag math -> prioritize `dim_left`.
This fails when `dim_left` is using dim tag math, but got some custom description from another layer, i.e. a case where we also would want to prioritize `dim_left`.

* If `dim_right` is undefined, `dim_left` is defined and its description is shorter than `dim_right` -> prioritize `dim_left`.
Very heuristically... Probably causes unexpected behavior.

* If `dim_right` is undefined, `dim_left` is defined and its base `Dim` object was created before `dim_right` -> prioritize `dim_left`.
This seems to behave as we want. And it seems simple. We can have an internal class-level creation counter which gets increased via `Dim.__init__`.

Or we don't add any special rule. But then the `out_dim` `description` will be ignored in many cases. Note for convolution etc, anything involving dim-tag-math: The description looks nice for one such computation, but once you have multiple, the expression gets very convoluted and long and hard to read, and a layer name or so would probably be more helpful to the user.

**Comment by albertz** [comment]
> If `dim_right` is undefined, `dim_left` is defined and its base `Dim` object was created before `dim_right` -> prioritize `dim_left`.
> This seems to behave as we want. And it seems simple. We can have an internal class-level creation counter which gets increased via `Dim.__init__`.

I implemented this now in https://github.com/rwth-i6/returnn/pull/1114.

**Comment by albertz** [comment]
All should be fixed now.

**Comment by albertz** [comment]
Hm, ok, the behavior now might still be somewhat confusing.
I had the intention to redefine the description of a new dim tag, and I used this code:
```python
spatial_dim.declare_same_as(nn.SpatialDim(f"{nn.NameCtx.current_ctx().get_abs_name()}:spatial"))
```
This does not work now, due to the fixes introduced here. In `declare_same_as`, this new rule applies:
```python
    if self.is_dim_known() and not other.is_dim_known():
      if self_same_as._creation_idx < other_same_base._creation_idx:
        # We want to keep self instead.
        # https://github.com/rwth-i6/returnn_common/issues/200
        other_same_base.declare_same_as(self_same_as)
        return
```
I think the introduced rule still makes sense, but maybe we should somehow adapt it or so.
In case that `spatial_dim` comes from extern data, the rule makes sense, and it should stay that way.
Maybe that should be the main distinction instead of `_creation_idx`?
Note that this is purely about cosmetic here. It's just about the description.
